lib/mix/test/mix/tasks/deps.git_test.exs

```diff
@@ -478,6 +478,204 @@ defmodule Mix.Tasks.DepsGitTest do
     purge([GitRepo, GitRepo.MixProject])
   end
 
+  describe "Git depth option" do
+    @describetag :git_depth
+
+    test "gets and updates Git repos with depth option" do
+      Process.put(:git_repo_opts, depth: 1)
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")})"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+
+        # Expand depth
+        update_dep(depth: 2)
+        Mix.Tasks.Deps.Get.run([])
+        assert_shallow("deps/git_repo", 2)
+
+        # Reduce depth
+        update_dep(depth: 1)
+        Mix.Tasks.Deps.Get.run([])
+        assert_shallow("deps/git_repo", 1)
+      end)
+    end
+
+    test "can be used with tag" do
+      Process.put(:git_repo_opts, depth: 1, tag: "with_module")
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")} - with_module)"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+      end)
+    end
+
+    test "can be used with branch" do
+      Process.put(:git_repo_opts, depth: 1, branch: "main")
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")} - main)"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+      end)
+    end
+
+    test "can be used with ref" do
+      [last, _ | _] = get_git_repo_revs("git_repo")
+
+      Process.put(:git_repo_opts, depth: 1, ref: last)
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")} - #{last})"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+      end)
+    end
+
+    test "changing refspec updates retaining depth" do
+      [last, first | _] = get_git_repo_revs("git_repo")
+
+      Process.put(:git_repo_opts, ref: first, depth: 1)
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")} - #{first})"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+        assert File.read!("mix.lock") =~ first
+
+        # Change refspec
+        update_dep(ref: last, depth: 1)
+        Mix.Tasks.Deps.Get.run([])
+        assert_shallow("deps/git_repo", 1)
+        assert File.read!("mix.lock") =~ last
+      end)
+    end
+
+    test "removing depth retains shallow repository" do
+      # The expectation is that one needs to deps.clean and deps.get the
+      # dependency for a full clone.
```

Comment:
Other people may not agree with this expectation. I went with it because there are different ways to "unshallow" a shallow clone, and what's available depends on the Git version (e.g. `--depth=large-number`, `--unshallow`) and are not equivalent in terms of performance and perhaps also of end result. Given the inherent complexity of interfacing with Git, I decided to keep the implementation as simple as I could imagine.

Reply:
There is nothing in `mix deps.get` that requires a deps.clean and deps.get and this would be a concern here. We should pick one of --depth or --unshallow, whatever is better, and require the user to be running on that Git version if they want to use depth.

Reply:
> There is nothing in `mix deps.get` that requires a deps.clean and deps.get and this would be a concern here.

I understand.

What would require a clean+get in my comment is to get a repo clone as if `:depth` had never been used on the dep. Other options also affect the repo clone in a permanent way that only clean+get can rectify. For example, using `:tag` causes tags to be fetched, but changing from `:tag` to a `:branch` will not "forget" the fetched tags. Removing `:branch` to go back to the main branch will not clean up the obsolete objects either.

 Are there documented guarantees on how exactly the clone is made and maintained over time?

---

Let's say we go with `--unshallow`.

The simpler implementation would pass that flag whenever `:depth` is not set. However, that would mean we'd impose a higher requirement on the Git version for everyone, even if they never used `:depth`.

The alternative involves trying to figure out if the repo clone is shallow and only then use `--unshallow`.

This extra complexity seemed unnecessary, as `depth` can be changed freely by the user, up and down, as well as changing tag/branch/ref will leave the repo in the desired state (with at least what has been required and no guarantees about extra objects like tags and additional branches or even orphaned refs).

Reply:
> What would require a clean+get in my comment is to get a repo clone as if :depth had never been used on the dep. 

Oh, then that's fine. :) The important is that it works. If you do a shallow and then non-shallow, it is expected that we won't automatically shallow it!

Reply:
> The important is that it works. If you do a shallow and then non-shallow, it is expected that we won't automatically shallow it!

Hmm do you want the behavior to match `git` operations or some other guiding principle?

I think we'll need to write down some examples of the successive states of the dep spec and expected final state to nail down the "it works" part :)

Then we can make sure the important behaviors are covered by tests.

Here are a few situations.

## Case A: removing `:depth`, same tag/branch/ref

1.
```elixir
    {:heroicons,
       github: "tailwindlabs/heroicons",
       tag: "v2.0.18",
       app: false,
       compile: false,
       sparse: "optimized",
       depth: 1}
```

| Property | Value |
| --- | --- |
| Working dir contents | tag `v2.0.18` |
| Number of commits reachable by `git log HEAD` | 1 |
| Other branches? | No guarantees |
| Other tags? | No guarantees |

2.
```elixir
    {:heroicons,
       github: "tailwindlabs/heroicons",
       tag: "v2.0.18",
       app: false,
       compile: false,
       sparse: "optimized"}
```

| Property | Value |
| --- | --- |
| Working dir contents | tag `v2.0.18` |
| Number of commits reachable by `git log HEAD` | _1 or N?_ * |
| Other branches? | No guarantees |
| Other tags? | No guarantees |

\* If `N`, then we need `--unshallow` if the tag didn't change, as `git fetch` will not automatically unshallow.

## Case B: adding `:depth``, same tag/branch/ref

1.
```elixir
    {:heroicons,
       github: "tailwindlabs/heroicons",
       tag: "v2.0.18",
       app: false,
       compile: false,
       sparse: "optimized"}
```

| Property | Value |
| --- | --- |
| Working dir contents | tag `v2.0.18` |
| Number of commits reachable by `git log HEAD` | N |
| Other branches? | No guarantees |
| Other tags? | No guarantees |

2.
```elixir
    {:heroicons,
       github: "tailwindlabs/heroicons",
       tag: "v2.0.18",
       app: false,
       compile: false,
       sparse: "optimized",
       depth: 1}
```

| Property | Value |
| --- | --- |
| Working dir contents | tag `v2.0.18` |
| Number of commits reachable by `git log HEAD` | 1 |
| Other branches? | No guarantees |
| Other tags? | No guarantees |

Reply:
We don't really care about other branches, tags, and commits inside `deps/foobar`. The important is that we can always fetch what is specified in the lock file or mix.exs. :)

Reply:
When `mix.exs` doesn't specify `:depth`, is there an expectation that the full history for the current branch/tag/ref is available?

Reply:
No, there shouldn’t be. All we guarantee is that we can migrate transparently whenever the lock or mix.exs changes.